crates/bevy_ecs/src/storage/thin_array_ptr.rs

```diff
@@ -0,0 +1,250 @@
+        let ptr = unsafe { self.data.as_ptr().add(index) };
+
+        // SAFETY:
+        // - The pointer is properly aligned
+        // - It is derefrancable (all in the same allocation)
+        // - `index` < `len` and the element is safe to write to, so its valid
+        // - We have a mutable reference to `self`
+        unsafe {
+            ptr.as_mut()
+                // SAFETY: We can use `unwarp_unchecked` because the pointer isn't null)
+                .unwrap_unchecked()
+        }
+    }
+
+    // TODO: Docs
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `index < len`
+    /// - ensure that `last_element_index` = `len - 1`
+    /// - update their saved length value to reflect that the last element has been removed (decrement it)
+    pub unsafe fn swap_remove_and_forget_unchecked(
+        &mut self,
+        index: usize,
+        last_element_index: usize,
+    ) -> *mut T {
+        if index != last_element_index {
+            std::ptr::swap_nonoverlapping(
+                self.get_unchecked_raw(index),
+                self.get_unchecked_raw(last_element_index),
+                1,
+            );
+        }
+        self.get_unchecked_raw(last_element_index)
+    }
+
+    // TODO: Docs
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `index < len`
+    /// - ensure that `last_element_index` = `len - 1`
+    /// - update their saved length value to reflect that the last element has been removed (decrement it)
+    pub unsafe fn swap_remove_and_drop_unchecked(
+        &mut self,
+        index: usize,
+        last_element_index: usize,
+    ) {
+        std::ptr::drop_in_place(self.swap_remove_and_forget_unchecked(index, last_element_index))
+    }
+
+    /// Pop the last element of the array.
+    ///
+    /// # Safety
+    /// - ensure that `current_len` is indeed the len of the array
+    /// - update their saved `len` (decrement it)
+    pub unsafe fn pop(&mut self, current_len: usize) -> Option<T> {
+        if current_len == 0 {
+            None
+        } else {
+            Some(std::ptr::read(self.data.as_ptr().add(current_len - 1)))
+        }
+    }
+
+    /// Clears the array, removing (and dropping) Note that this method has no effect on the allocated capacity of the vector.
+    ///
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `current_len` is indeed the length of the array
+    /// - update their saved length value
+    pub unsafe fn clear_elements(&mut self, mut current_len: usize) {
+        while self.pop(current_len).is_some() {
+            current_len -= 1;
+        }
+    }
+
+    /// Drop the entire array and all its elements.
+    ///
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `current_len` is indeed the length of the array
+    /// - ensure that `current_capacity` is indeed the capacity of the array
```

Comment:
The caller must also make sure not to use the ThinArray in any way after calling this.